src/main/java/com/myorg/mvc/service/UserServiceImpl.java

```diff
@@ -0,0 +1,117 @@
+        Optional<User> userByEmail = userRepository.findByEmail(userToSave.getEmail());
+        if (userByEmail.isPresent()) {
+            throw new DoubleRegistrationException(userToSave.getEmail());
+        }
+        User userToInsert = createEmptyUser();
+        userToInsert.setFirstName(userToSave.getFirstName());
+        userToInsert.setLastName(userToSave.getLastName());
+        userToInsert.setEmail(userToSave.getEmail());
+        userToInsert.setBirthday(userToSave.getBirthday());
+        String userName = (userToSave.getFirstName() + userToSave.getLastName()).toLowerCase(Locale.ROOT);
+        userToInsert.setUsername(userName);
+        userToInsert.setPassword(passwordEncoder.encode(userToSave.getPassword()));
+
+        return userRepository.save(userToInsert);
+    }
+
+    public User deleteUser(Long userId) {
+        User userToDelete = userRepository.findById(userId).orElseThrow(() -> new UserNotFoundException(userId));
+        userRepository.deleteById(userId);
+
+        return userToDelete;
+    }
+
+    public User updateUser(Long userId, String firstName, String lastName, String email, String birthday) {
+        User user = userRepository.findById(userId).orElseThrow(() -> new UserNotFoundException(userId));
+
+        String userEmail = user.getEmail();
+        insertValues(user, userEmail, email, firstName, lastName, birthday);
+        return user;
+    }
+
+
+    @Override
+    public User createEmptyUser() {
+        return new User();
+    }
+
+    @Override
+    public User updateUser(Long userId, User user) {
+        User repoUser = userRepository.findById(userId).orElseThrow(() -> new UserNotFoundException(userId));
+        String repoUserEmail = repoUser.getEmail();
+        String userEmail = user.getEmail();
+        String firstName = user.getFirstName();
+        String lastName = user.getLastName();
+        String userBirthday = user.getBirthday();
+
+        insertValues(repoUser, repoUserEmail, userEmail, firstName, lastName, userBirthday);
+        return user;
+    }
+
+    private void insertValues(User repoUser, String repoUserEmail, String userEmail, String firstName, String lastName, String userBirthday) {
+        if (isValidName(firstName, lastName)) {
+            repoUser.setFirstName(firstName);
+            repoUser.setLastName(lastName);
+        }
+
+        if (isValidEmail(userEmail, repoUserEmail)) {
+            repoUser.setEmail(userEmail);
+        } else throw new DoubleRegistrationException(repoUser.getEmail());
+
+        repoUser.setBirthday(userBirthday);
+    }
+
+    private boolean isValidEmail(String emailAddress, String userEmail) {
+        return !Objects.equals(userEmail, emailAddress) && Pattern.compile(UserServiceImpl.REGEX_PATTERN).matcher(emailAddress).matches();
+    }
+
+    private boolean isValidName(String firstName, String lastName) {
+        return firstName != null && lastName != null && firstName.length() > 0 && lastName.length() > 0;
```

Comment:
1 Format it
```
return firstName != null
                && lastName != null
                && firstName.length() > 0
                && lastName.length() > 0;
```
2.
firstName.length() > 0 can be replaced to !firstName.trim().isEmpty()